### ISSUE TITLE
Occasional KL loss spike with FSDP workers in grpo flow

### ISSUE THREAD

**Issue by hchings** [verl-project/verl#5274]
### System Info

----------Python Info----------
Version      : 3.12.3
Compiler     : GCC 13.3.0
Build        : ('main', 'Aug 14 2025 17:47:21')
Arch         : ('64bit', 'ELF')
------------Pip Info-----------
Version      : 25.2
Directory    : /usr/local/lib/python3.12/dist-packages/pip
vllm         : not found.
sglang       : not found.
ray          : 2.53.0
torch        : 2.9.0a0+145a3a7bda.nv25.10
----------verl Info-----------
Version      : 0.8.0.dev
Directory    : /scratch/verl-integ/verl
Commit Hash  : 0758297e2c1d664c2524cbd65f739c43e65ccf51
----------Platform Info----------
Platform     : Linux-6.8.0-90-generic-x86_64-with-glibc2.39
system       : Linux
node         : e6522036bd27
release      : 6.8.0-90-generic
version      : #91-Ubuntu SMP PREEMPT_DYNAMIC Tue Nov 18 14:14:30 UTC 2025
----------Environment----------
CUDA Runtime : 13.0
CUDA Compiler : Cuda compilation tools, release 13.0, V13.0.88
----------System Info----------
GPU Count       : 8
GPU   Type    : NVIDIA H100

### Issue description
<img width="976" height="550" alt="Image" src="https://github.com/user-attachments/assets/1bac9d1d-4261-467a-954b-3e9bb6b482ce" />

We noticed that the kl loss of commit `ded57b0b` + [MR5212](https://github.com/verl-project/verl/pull/5212/commits) (pink line above) is consistently higher than without MR5212 (which is a simple clean cache). Originally, we thought the clean cache and `gc.collect` in the trainer process triggered some implicit behavior in either trainer or rollout. However, when we rebased to  `ad404cb5`, there's no longer such issue. 

This issue currently does not block us nor the users, since with latest main branch, the kl loss is normal and matches that of vllm. But given that we still occasionally see it with the same verl main branch and config yet slightly different ray submitter, we still open this Issue to track this. 

- This issue only happens with FSDP workers + TRTLLM rollout.  Megatron + TRTLLM has no such issue, neither does FSDP + vLLM rollout. 
- We've ruled out several potential issues from TRTLLM rollout side (e.g., weight update between trainer and rollout processes are correct), and haven't found the root cause in FSDP trainer. 
- We've ruled out that the diff was caused by #5203.
- Per discussion w/ verl team (Xibin Wu), the gsm8k dataset used in grpo example is too easy to identify problems, and we were recommended to do ablation on the DAPO dataset instead. With DAPO aime-2024, we don't see issues in `acc/mean` curve between the two commits mentioned in Repro section.

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Checkout commit `ded57b0b` from verl repo main branch
2. Run script and log curves with wandb: `bash examples/grpo_trainer/run_qwen2-7b_math_trtllm.sh` (run til at lease `50` steps for the issue to show)
3. Do the same but with commit `ad404cb5`, and compare the KL loss curves between these two.

### Expected behavior

FSDP worker to have a stable kl-loss given same datasets and config.

**Comment by Kirrito-k423** [comment]
This is an interesting issue. The KL loss spike with FSDP + TRTLLM but not with Megatron + TRTLLM or FSDP + vLLM suggests a subtle difference in how weight synchronization or KV cache handling works between the combinations.

Key areas to investigate:

1. **Weight sync timing**: Between commit `ded57b0b` and `ad404cb5`, there may have been changes to when weights are synchronized from trainer to rollout workers. Early weight sync could cause stale weights during generation.

2. **TRTLLM KV cache reuse**: TRTLLM may be reusing KV cache across steps when it shouldn't. The `gc.collect` and cache clearing may be affecting this behavior.

3. **FSDP state_dict behavior**: FSDP's `state_dict` call has different behavior depending on `state_dict_type` and whether parameters are offloaded. Changes between commits may have affected this.

**Debugging steps:**

1. Add logging for weight hash before/after sync:
```python
weight_hash = sum(p.sum().item() for p in model.parameters())
print(f"[DEBUG] Weight hash before rollout: {weight_hash}")
```

2. Check TRTLLM's KV cache clearing:
```python
# Ensure cache is cleared between steps
actor_rollout_ref.rollout.free_cache_engine=True
```

3. Compare weight transfer implementations between FSDP and Megatron paths.

This intermittent behavior suggests a race condition or non-deterministic ordering issue. Consider filing a more detailed repro if the issue persists on newer commits.

**Comment by Kirrito-k423** [comment]
中文：感谢记录这个间歇性问题。我重新核对了当前公开证据后，暂时建议先补充最新 main 上的可复现信息，而不是直接关闭或归因为 #5212。

目前能确认的证据是：Issue 本身说明 latest main 上 KL loss 正常并匹配 vLLM，但在相同 main/config、不同 Ray submitter 下偶尔仍可见；#5212 确实只是在 TRTLLM `update_weight()` 后增加缓存清理以降低峰值显存；当前 main 的 TRTLLM rollout 仍在权重 IPC handle 更新、finalize update_weights、barrier 后执行 `gc.collect()` / `empty_cache()`，相关路径是 `verl/workers/rollout/trtllm_rollout/trtllm_rollout.py`；当前 main 的 FSDP 权重导出路径在 `verl/workers/engine/fsdp/transformer_impl.py` 中已有 FSDP1/FSDP2 staging、offload 和 DTensor materialization 处理；当前 main 已没有复现步骤里的 `examples/grpo_trainer/run_qwen2-7b_math_trtllm.sh`，文档中的 TRTLLM+FSDP GRPO 示例现在是 `INFER_BACKEND=trtllm bash examples/grpo_trainer/run_qwen3_8b_fsdp.sh`，见 `docs/workers/trtllm_worker.rst`。

为了继续定位，能否补充一次在当前 main 上仍能复现 spike 的最小信息？建议包括：1. 具体 verl commit hash；2. 完整启动脚本/config diff；3. 两种 Ray submitter 的命令和关键环境变量差异；4. 对应 wandb 曲线链接或 step 范围；5. 是否使用当前文档里的 Qwen3 TRTLLM+FSDP 路径也能复现。

English: Thanks for tracking this intermittent behavior. After re-checking the public evidence, I think this needs a current-main repro package before we can safely close it or attribute it to #5212.

What is currently supported by evidence: the issue says latest main has normal KL loss matching vLLM, while the spike may still appear occasionally with the same main/config but a different Ray submitter; #5212 did add cache cleanup after TRTLLM `update_weight()` to reduce peak memory; current main still finalizes TRTLLM weight updates and then runs `gc.collect()` / `empty_cache()` in `verl/workers/rollout/trtllm_rollout/trtllm_rollout.py`; current main's FSDP export path in `verl/workers/engine/fsdp/transformer_impl.py` includes FSDP1/FSDP2 staging, offload, and DTensor materialization handling; the original repro script `examples/grpo_trainer/run_qwen2-7b_math_trtllm.sh` is no longer present on current main, and the documented TRTLLM+FSDP GRPO path is now `INFER_BACKEND=trtllm bash examples/grpo_trainer/run_qwen3_8b_fsdp.sh` in `docs/workers/trtllm_worker.rst`.

Could you provide a minimal current-main repro when the spike still happens? The most useful details would be: 1. exact verl commit hash; 2. full launch script/config diff; 3. the two Ray submitter commands and key environment differences; 4. wandb links or affected step range; 5. whether the current documented Qwen3 TRTLLM+FSDP path also reproduces it.
This response was generated by a scheduled Codex GPT5.5 issue processing task.
